### PR TITLE
[HZ-758] Replace the json-path library with one that supports SQL standard

### LINKED ISSUE #19821: Replace the json-path library with one that supports SQL standard [HZ-758]
Json-path wasn't standardized, though it was widely used. When SQL incorporated it, they actually formally specified it, but they did some changes, for example:
- array ranges:
  - sql: `$[0 to 2]`
  - hz-sql currently: `$[0-2]`
- quoted properties:
  - sql: `$."foo prop"`
  - hz-sql currently: `$['foo prop']

SQL also supports `strict` and `lax` modes, which are enabled using modifiers at the expression start. Strict mode throws error on out-of-range array indexes or non-existent properties.

We need to fix this and support SQL standard syntax. Whether to support non-sql-standard syntax that's common elsewhere is an open question. We can do it by modifying the json-surfer lib, which also has other benefits:
- uses a parser generator, so it's easier to modify
- works in a streaming way - we don't need to unmarshall the whole JSON object into memory
- supports different formats, it can be used to implement binary json format in the future.

On the other side, it's less widely used, so it can have child issues.

### PR DESCRIPTION
**Pull request by galibey**
Fixes #19821

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible

### CODE REVIEW

**Review by ivanthescientist — APPROVED** [review]

**Comment by viliam-durina** [comment]
Changes done with my commits will pass after hazelcast/JsonSurfer#25 is merged.

File: hazelcast-sql/pom.xml
Comment (frant-hartm) [inline hazelcast-sql/pom.xml:179]: This is removed, so it should be removed from the relocations section below.

File: hazelcast-sql/pom.xml
Comment (frant-hartm) [inline hazelcast-sql/pom.xml:181]: Do you want to relocate the antlr under `com.hazelcast.org.antlr` or similar? It's done for all 3rd party deps in sql module.

File: pom.xml
Comment (frant-hartm) [reply pom.xml:1657]: This is in the `dependencyManagement` section. Usually, when multiple modules (e.g. jet extensions) depend transitively on different versions of a library, we put the library here with a pinned version so we end up with a single version of the class files on classpath.

I think in this case you want to relocate antlr in sql module and remove it from here.

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast-sql/pom.xml
Comment (ivanthescientist) [reply hazelcast-sql/pom.xml:179]: Sadly we can't remove it actually, its part of Calcite's core dependencies so it has to be still shaded in.

File: hazelcast-sql/pom.xml
Comment (ivanthescientist) [reply hazelcast-sql/pom.xml:183]: Should be shaded now.

**Comment by viliam-durina** [comment]
run-lab-run

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunctionIntegrationTest.java
Comment (ivanthescientist) [inline hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunctionIntegrationTest.java:110]: Think I missed this, but `JSON_QUERY` is not supposed to ever return a primitive (non-Object/Array) value. 
With the default `WRAPPER` value (`WITH CONDITIONAL WRAPPER` iirc) its supposed to return `[3]` here.

**Review by ivanthescientist — CHANGES_REQUESTED** [review]

File: hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunctionIntegrationTest.java
Comment (viliam-durina) [reply hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunctionIntegrationTest.java:110]: Hmm, I double checked the spec, and indeed, it's not supposed to return a JSON literal. However, Oracle supports it and IMO, it makes perfect sense. A standalone JSON literal is a well-formed JSON document:
image

`JSON_QUERY` returns a json documents. IMO the restriction that the JSON document can be only json-null, array and object is arbitrary. Why exclude literals?

The difference between JSON_QUERY and JSON_VALUE on literals is also illustrated here:
image

I'm for keeping it. It's an extension of the standard, but one that IMO should have been there from the beginning. Maybe it's even in some newer release...

**Review by ivanthescientist — APPROVED** [review]

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunction.java
Comment (viliam-durina) [inline hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunction.java:111]: It deserves a comment why we, contrary to the usual practice, we don't chain the cause.

File: hazelcast-sql/pom.xml
Comment (galibey) [reply hazelcast-sql/pom.xml:183]: `hazelcast-sql` depends on `JsonSurfer` which depends on `antlr-runtime`. Will classes from `JsonSurfer` be aware of such relocation?

File: pom.xml
Comment (galibey) [reply pom.xml:1652]: Before this PR, there was one test dependency of `antlr` in `hazelcast` and one or two compile dependencies in `hazelcast-jet-extensions`, bringing one more to `hazelcast-sql` fails build by maven-enforcer-plugin but this change won't be needed once shading and relocating will be done properly.

File: hazelcast-sql/pom.xml
Comment (galibey) [reply hazelcast-sql/pom.xml:183]: I checked, works as expected. `JsonSurfer` classes use shaded `antlr-runtime`.

File: hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunction.java
Comment (galibey) [reply hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonQueryFunction.java:114]: Added comments here

**Review by viliam-durina — APPROVED** [review]
From my side it's OK now, but let's still wait for @frant-hartm's approval.

**Review by frant-hartm — APPROVED** [review]
LGTM now, thanks for the updates!!

**Pull request merged**
